activerecord/lib/active_record/attributes.rb

```diff
@@ -260,14 +237,25 @@ def define_attribute(
         define_default_attribute(name, default, cast_type, from_user: user_provided_default)
       end
 
-      def load_schema! # :nodoc:
-        super
-        attributes_to_define_after_schema_loads.each do |name, (cast_type, default)|
-          cast_type = cast_type[type_for_attribute(name)] if Proc === cast_type
-          define_attribute(name, cast_type, default: default)
+      def _default_attributes # :nodoc:
+        @default_attributes ||= begin
+          attributes_hash = columns_hash.transform_values do |column|
+            ActiveModel::Attribute.from_database(column.name, column.default, type_for_column(column))
+          end
+
+          attribute_set = ActiveModel::AttributeSet.new(attributes_hash)
+          apply_pending_attribute_modifications(attribute_set)
+          attribute_set
         end
       end
```

Comment:
With this, `@default_attributes` is now computed atomically.  `@columns_hash` is also [computed atomically](https://github.com/rails/rails/blob/53438e9216243fb75400c8800ddf549daa52eab3/activerecord/lib/active_record/model_schema.rb#L616-L618).  It might be feasible to eventually remove the monitor from `load_schema`.